File: javascript/robot-name/robot-name.js
// This is only a SKELETON file for the 'Robot Name' exercise. It's been
// provided as a convenience to get your started writing code faster.

let robotNames = new Set()
const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
const DIGITS = "0123456789"

let randomName = () => {
    const ALPHA_CHARS = ALPHABET.length
    const DIGIT_CHARS = DIGITS.length
    let retry = true
    let name = ''
    while (retry)
    {
        name = ALPHABET[Math.floor(Math.random() * ALPHA_CHARS)] + 
            ALPHABET[Math.floor(Math.random() * ALPHA_CHARS)] +
            DIGITS[Math.floor(Math.random() * DIGIT_CHARS)] + 
            DIGITS[Math.floor(Math.random() * DIGIT_CHARS)] + 
            DIGITS[Math.floor(Math.random() * DIGIT_CHARS)]
        retry = robotNames.has(name)
    }
    robotNames.add(name)
    return name
}

export class Robot {
    constructor()
    {
        this._name = ""
        this.reset()
    }

    reset()
    {
        // I think you should release a name when reset
        // but the tests say otherwise.

        //if (this.name !== "")
        //    robotNames.delete(this.name)
        
        this._name = randomName()
    }

    get name()
    {
        return this._name
    }
}



Robot.releaseNames = () => {robotNames.clear()};

File: go/robot-name/robot_name.go
// Robot naming exercise.
package robotname

import (
	"errors"
	"math/rand/v2"
)

const MAX_NAMES = 26 * 26 * 10 * 10 * 10 // 676,000

var all_robot_names map[string]bool = make(map[string]bool)

// Define the Robot type here.
type Robot struct {
	name string
}

// Return the name of the robot, generating a new name if none has been asssigned yet.
// @returns: The existing/new robot name.
// @raises: Error if there is no more namespace to allocate robot names (676,000 maximum).
func (r *Robot) Name() (string, error) {
	const ALPHABET_LEN = 26
	const DIGIT_LEN = 10
	if r.name == "" {
		if len(all_robot_names) >= MAX_NAMES {
			return "", errors.New("All robot names have been allocated.")
		}
		name_ok := false
		robot_name := ""
		for !name_ok {
			robot_name = ""
			robot_name += string(rune('A' + rand.IntN(ALPHABET_LEN)))
			robot_name += string(rune('A' + rand.IntN(ALPHABET_LEN)))
			robot_name += string(rune('0' + rand.IntN(DIGIT_LEN)))
			robot_name += string(rune('0' + rand.IntN(DIGIT_LEN)))
			robot_name += string(rune('0' + rand.IntN(DIGIT_LEN)))
			_, name_ok = all_robot_names[robot_name]
			name_ok = !name_ok
		}
		all_robot_names[robot_name] = true
		r.name = robot_name
	}
	return r.name, nil
}

// Reset the robot blanking its name. I think that should allow reuse of the now unused name but the tests say otherwise.
func (r *Robot) Reset() {
	r.name = ""
}
